src/V3Funcopt.cpp

```diff
@@ -0,0 +1,171 @@
+// -*- mode: C++; c-file-style: "cc-mode" -*-
+//*************************************************************************
+// DESCRIPTION: Verilator: Generic optimizations on a per function basis
+//
+// Code available from: https://verilator.org
+//
+//*************************************************************************
+//
+// Copyright 2003-2024 by Wilson Snyder. This program is free software; you
+// can redistribute it and/or modify it under the terms of either the GNU
+// Lesser General Public License Version 3 or the Perl Artistic License
+// Version 2.0.
+// SPDX-License-Identifier: LGPL-3.0-only OR Artistic-2.0
+//
+//*************************************************************************
+//
+// - Split assignments to wide locations with Concat on the RHS
+//   at word boundaries:
+//    foo = {l, r};
+//   becomes (recursively):
+//    foo[_:_] = r;
+//    foo[_:_] = l;
+//
+//*************************************************************************
+
+#include "V3PchAstMT.h"
+
+#include "V3Funcopt.h"
+
+#include "V3Global.h"
+#include "V3Stats.h"
+#include "V3ThreadPool.h"
+
+VL_DEFINE_DEBUG_FUNCTIONS;
+
+class FuncoptVisitor final : public VNVisitor {
+    // NODE STATE
+    //  AstNodeAssign::user()     -> bool.  Already checked, safe to split. Omit expensive check.
+
+    // STATE
+    AstCFunc* const m_funcp;  // The function being processes
+
+    VDouble0 m_concatSplits;  // Number of splits in assignments with Concat on RHS
+
+    static bool rhsReadsLhs(AstNodeAssign* nodep) {
```

Comment:
Although strange, you can have the lhs read lhs.

`foo[foo[1:0]] = bar`

I suspect other things will break too if we test this, but might as well disable this opt if this is found, and avoid adding a new one ;)

Reply:
Agreed.

That predicate is pretty much copy paste from V3Premit, just FYI.

We should have an early pass that serializes these conflicts in a deterministic way, you can also have crazy thing like `foo[dpi(outparam, inparam++)] = outparam ^ inparam;`

For now I'm happy for a mushroom cloud to appear when the user did that though.